### DISCUSSION TITLE
What to do on version incompatibilities between IDE and the project's specified engine

### DISCUSSION THREAD

**Discussion by hubertp** [enso-org/enso#7639] [Work In Progress - Engine]
We had a number of issues recently related to the fact that newer IDE was forced to use incompatible engine version.
Examples include:
- https://github.com/enso-org/enso/issues/7095
- https://github.com/enso-org/enso/issues/7426
and numerous discussions about the version of newly created projects having a _sticky_ and _old_ version of the engine, therefore causing immediate problems.

While it would be nice to be always compatible between IDE and engine I don't think it is realistic at this point in time.
Specifying which versions break the protocol and therefore should be upgraded is also non-trivial and rather error prone (easy to miss the changes). I've seen projects defining such logic but it is a massive can of worms.

Should we for now abandon the whole fixed version schema and always use latest?

**Comment by radeusgd** [comment]
> Should we for now abandon the whole fixed version schema and always use latest?

I think indeed using latest seems like a good idea. Still IMHO it may be worth to keep the `edition` in `package.yaml` by default marking the version with which the project was first created. 

Once the project is opened in a different IDE/engine version combination, we could allow the different version but display a warning note to the user saying that "the project was for a different Enso version so there may be some API incompatibilities" (similar to the current "Unsupported engine version" message).

**Reply by JaroslavTulach** [reply]
What is a difference between:
- "Unsupported engine version"?
- "the project was for a different Enso version so there may be some API incompatibilities"?

If the project is for different engine version, then we have an unsupported engine version, don't we?

**Reply by radeusgd** [reply]
The latter is about project being incompatible with Enso version.

The former - "Unsupported engine version" was a message that was supposed to indicate that the IDE and engine versions are not compatible with each other.

**Reply by JaroslavTulach** [reply]
I see. So we have three elements here:
- the version of the engine requested by project inside of its `package.yaml`
- the version of the engine started by `project-manager`
- the version of the engine the IDE wants to work with

**Comment by hubertp** [comment]
One problem with using "latest" is that "latest" is not really a version. And I would expect a different set of problems caused by that fact and it would also make it harder for us to debug.
Other thoughts:
- display engine version number next to the project name or in some project's description tab (we don't have such a thing yet)
- maybe IDE could keep track of the version of the engine, compare it to the version of the IDE and **iff** there are problems in communication/execution display a warning that user should consider bumping the version

I particularly like the second one.

**Reply by JaroslavTulach** [reply]
Checking whether the IDE can communicate with the engine/LS process is indeed important. Doesn't that imply we should version the protocol? Shouldn't [connection management](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md#connection-management) messages specify version?

Or how else the IDE and engine/LS negotiate their versions and display of "Unsupported engine version"?

I see bunch of [capabilities](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md#capability-management) whose could also be used for some _form of versioning_. Anyway its all about attitude:

If we care about compatibility, then let us say: we are _not going to do any incompatible changes_ to the protocol since today (tomorrow or ...)!

**Comment by vitvakatu** [comment]
> iff there are problems in communication/execution display a warning that the user should consider bumping the version

I believe detecting such issues is hard, and distinguishing between incompatibilities and bugs is even more complicated. 

We should split this problem into two parts:
1. What behavior is better for us while developing the project?
2. What would be better for our users, especially considering the upcoming release?

I think the answer for `2` includes the current behavior when the user will use the engine the project was created with. Therefore, they won't need to rewrite the code each time to accommodate changes in the standard library. At the same time, it should be _our_ job to keep track of breaking changes between the IDE and the engine and warn users about it.

> Once the project is opened in a different IDE/engine version combination, we could allow the different version but display a warning note to the user saying that "the project was for a different Enso version so there may be some API incompatibilities" (similar to the current "Unsupported engine version" message).

I don't like this approach because _there will be_ incompatibilities, and the user will be forced to rewrite the project instead of doing his job. If I open my project from last week and see that it is not working with the new IDE __and I can't change that because the latest engine is always used__, I will be infuriated.

**Reply by JaroslavTulach** [reply]
> the user will use the engine the project was created with

In the context of [this classification](https://github.com/orgs/enso-org/discussions/7639#discussioncomment-6839016), this approach fixes the mismatch between _project and engine_. But it doesn't address the incompatibilities between the _IDE and engine/LS_.

There is no guarantee new IDE will understand the [old protocol](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md). Moreover there is one more problem...

**Reply by JaroslavTulach** [reply]
New **visualizations** (which are currently part of the IDE) and **drop down support** are not guaranteed to work with an old engine! We shall _version also the visualizations_. The JavaScript parts of visualization shall _check for capabilities_ of the `Standard.Visualization` module, shall avoid using newer features when running with old engine & co. Complex, but anyway its all about attitude:

If we care about compatibility, then let us say: we are going to _write our JavaScript visualizations defensively_, ready to work with older `Standard.Visualization` version since today (tomorrow or ...)!

**Comment by 4e6** [comment]
> If I open my project from last week and see that it is not working with the new IDE and I can't change that because the latest engine is always used, I will be infuriated.

We're still in the early stage of the project, shipping many features and bug fixes with every release. And currently, the situation is the opposite. When you download a new IDE, you can't use it with the old (pinned) engine version because of the changes made since.

First, if we want to have versioning, we should ensure that the user has a clear understanding of it, i.e.
- Make the user aware that the project they created has a version (currently they are not)
- If the project version is outdated, make the user aware of it and allow them to change the version (at least to the version of a new IDE they have downloaded) 

Then, we should distinguish between the engine version and the language server version. 
- The engine version includes the compiler and the std-lib. This is the version that goes to the `project.yaml` and which the user can control.
- The language server API version should be bound to the IDE version as they are shipped together and work together as one system.  The language server API should be always new as it is shipped together with the new IDE release.

Having separate engine and language server versions can cause a situation when a new language server version will not support older engine versions. But that's ok, and instead of failing to open the project or experiencing random issues in IDE (which is happening now), the user will be aware of it and will be forced to switch to a new engine version (or to use older IDE versions).

**Reply by JaroslavTulach** [reply]
> Having separate engine and language server versions can cause a situation when a new language server version will not support older engine versions.

How am I supposed to interpret this sentence, @4e6? Our build creates following distribution:
```
enso$ ./built-distribution/enso-engine-0.0.0-dev-linux-amd64/
```
that distribution contains engine _as well as_ language server implementation. E.g. engine version and language server implementation are tight together.

> language server API version should be bound to the IDE version as they are shipped together

_Language server API version_ should be a number specified inside of the [specification document](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md). 
- then there is _language server implementation_ (which also may have a _version_) which resides inside of `enso-engine` distribution and implements some version of the _language server API_
- the IDE just uses the _language server API_ and shall specify the _lowest API version_ it requires to function properly - IDE shall reject to work with engine/LS older than the _lowest API version_.
 
However the IDE will also work with all newer versions of the _language server API_, while they are marked as _compatible with the lowest API version_.

**Comment by 4e6** [comment]
To summarize discussions that happened since the last message.

- instead of versioning the language server API, we want to always use the latest IDE + the latest language server
- language server will keep track of breaking changes of the internal engine API, and maintain the oldest supported engine version (edition)
- I'm not sure if editions are comparable (they are definitely not semantic). I assume that we don't want to touch the edition, so instead we can introduce a private semantic version that comes together with edition. This private version will be used by the language server to determine if the engine is supported or not.

**Reply by JaroslavTulach** [reply]
How will you identify "latest language server" when there is no versioning?

**Reply by JaroslavTulach** [reply]
How can "language server will keep track of breaking changes" when there is no version to begin with?

**Reply by 4e6** [reply]
We can use the one that is shipped with IDE.
